services/product.js

```
@@ -0,0 +1,171 @@
+        //     },
+        //     {
+        //         "productId": "Prmxic",
+        //         "productName": "Product26",
+        //         "productOwnerName": "Anthony Shanthan",
+        //         "Developers": [
+        //             "Phoenix Gutierrez",
+        //             "James Andreas",
+        //             "Adam Salyer"
+        //         ],
+        //         "scrumMasterName": "Aidan McGhie",
+        //         "startDate": "2022/11/30",
+        //         "methodology": "Agile"
+        //     },
+        // ];
+
+        // await saveProducts()
+        // await loadProducts();
+    } else {
+        // await loadProducts();
+    }
+    return products 
+}
+
+async function getById(id) { 
+    const itemFound = (await getAll()).find((product) => product.productId === id)
+    if (!itemFound) {
+        throw new Error ("Product was not found!");
+    } else {
+        return itemFound;
+    }
+}
+
+async function update(product) { 
+    const idxProductFound = await getById(product.productId)
+    idxProductFound.productName = product.productName
+    idxProductFound.productOwnerName = product.productOwnerName
+    idxProductFound.Developers = product.Developers
+    idxProductFound.scrumMasterName = product.scrumMasterName
+    idxProductFound.startDate = product.startDate
+    idxProductFound.methodology = product.methodology
+    
+    console.log("update: products", products)
+    await saveProducts()
+}
+
+async function create(product) { 
+    const id = generateUId()
+    const productName = product.productName ? product.productName : ""
+    const productOwnerName = product.productOwnerName ? product.productOwnerName : ""
+    const developers = product.Developers ? product.Developers : []
+    const scrumMasterName = product.scrumMasterName ? product.scrumMasterName : ""
+    const startDate = product.startDate ? product.startDate : ""
+    const methodology = product.methodology ? product.methodology : ""
```

Do you really want  all these variables to default to empty strings and arrays? Maybe it's better to throw an error if something is missing.

---

done.
Just a thought: Do we need to check for errors in input in both the helper function and the user function?  Is there any way to simplify the code so I am not repeating the logic in both functions?

---

No, the `createProduct` function should be called `validateProduct` and you can re-use this new validation function everywhere you need it

---

Done - validateProduct function is created for validation of product, so checking errors is only done in one place.